applications/snake-game/snake-game.c

```diff
@@ -0,0 +1,417 @@
+            snake_state->currentMovement = right;
+            break;
+        case left:
+            snake_state->currentMovement = left;
+        default:
+            break;
+        }
+        break;
+    case left:
+        switch(snake_state->nextMovement) {
+        case up:
+            snake_state->currentMovement = up;
+            break;
+        case down:
+            snake_state->currentMovement = down;
+        default:
+            break;
+        }
+        break;
+    }
+}
+
+Point get_next_step(SnakeState const *const snake_state) {
+    Point next_step = snake_state->points[0];
+    switch(snake_state->currentMovement) {
+    // +-----x
+    // |
+    // |
+    // y
+    case up:
+        next_step.y--;
+        break;
+    case right:
+        next_step.x++;
+        break;
+    case down:
+        next_step.y++;
+        break;
+    case left:
+        next_step.x--;
+        break;
+    }
+    return next_step;
+}
+
+void move_snake(SnakeState *const snake_state, Point const next_step) {
+    memmove(
+        snake_state->points + 1,
+        snake_state->points,
+        snake_state->len * sizeof(*(snake_state->points)));
+    snake_state->points[0] = next_step;
+}
+
+void process_game_step(SnakeState *const snake_state) {
+    if(snake_state->state == gameOver) {
+        return;
+    }
+
+    bool can_turn = (snake_state->points[0].x % 2 == 0) && (snake_state->points[0].y % 2 == 0);
+    if(can_turn) {
+        turn_snake(snake_state);
+    }
+
+    Point next_step = get_next_step(snake_state);
+
+    bool crush = collision_with_frame(next_step);
+    if(crush) {
+        if (snake_state->state == life) {
+            snake_state->state = lastChance;
+            return;
+        } else if (snake_state->state == lastChance) {
+            snake_state->state = gameOver;
+            return;
+        }
+    } else {
+        if (snake_state->state == lastChance) {
+            snake_state->state = life;
+        }
+    }
+
+    crush = collision_with_tail(snake_state, next_step);
+    if(crush) {
+        snake_state->state = gameOver;
+        return;
+    }
+
+    bool eatFruit = (next_step.x == snake_state->fruit.x) && (next_step.y == snake_state->fruit.y);
+    if(eatFruit) {
+        snake_state->len++;
+        if(snake_state->len >= MAX_SNAKE_LEN) {
+            snake_state->state = gameOver;
+            return;
+        }
+    }
+
+    move_snake(snake_state, next_step);
+
+    if(eatFruit) {
+        place_new_fruit(snake_state);
+    }
+}
+
+void game_logic_thread(void* ctx) {
+    for(;;) {
+        delay(200);
+
+        SnakeState* snake_state = acquire_mutex((ValueMutex*)ctx, 25);
+        if(snake_state == NULL) {
+            continue;
+        }
+        process_game_step(snake_state);
+        release_mutex((ValueMutex*)ctx, snake_state);
+    }
+}
+
+int32_t snake_game_app(void* p) {
+    osMessageQueueId_t event_queue = osMessageQueueNew(8, sizeof(InputEvent), NULL);
+
+    SnakeState snake_state;
```

Comment:
alloc is preferred way to allocate memory: easier to grow structures in small stack runtime.